stellargraph/layer/graph_classification.py

```diff
@@ -0,0 +1,236 @@
+                node features (size batch_size x N x F),
+                graph adjacency matrix (batch_size x N x N),
+                where N is the number of nodes in the graph, and F is the dimensionality of node features.
+
+        Returns:
+            Keras Tensor that represents the output of the layer.
+        """
+        features, A = inputs
+
+        h_graph = K.batch_dot(A, features)
+        output = K.dot(h_graph, self.kernel)
+
+        # Add optional bias & apply activation
+        if self.bias is not None:
+            output += self.bias
+        output = self.activation(output)
+
+        return output
+
+
+@experimental(reason="Missing unit tests")
+class GraphClassification:
+    """
+    A stack of :class:`GraphClassificationConvolution` layers together with a Keras `GlobalAveragePooling1D` layer
+    that implement a supervised graph classification network using the GCN convolution operator
+    (https://arxiv.org/abs/1609.02907).
+
+    The model minimally requires specification of the GCN layer sizes as a list of ints
+    corresponding to the feature dimensions for each hidden layer,
+    activation functions for each hidden layers, and a generator object.
+
+    To use this class as a Keras model, the features and pre-processed adjacency matrix
+    should be supplied using the :class:`GraphGenerator` class.
+
+    Notes:
+      - The inputs are tensors provided by the :class:`GraphGenerator` object.
+
+    Examples:
+        Creating a graph classification model from a list of :class:`StellarGraph`
+        objects (``graphs``). We also add two fully connected dense layers using the last one for binary classification
+        with `softmax` activation::
+
+            generator = GraphGenerator(graphs)
+            model = GraphClassification(
+                             layer_sizes=[32, 32],
+                             activations=["elu","elu"],
+                             generator=generator,
+                             dropout=0.5
+                )
+            x_inp, x_out = model.build()
+            predictions = Dense(units=8, activation='relu')(x_out)
+            predictions = Dense(units=2, activation='softmax')(predictions)
+
+    Args:
+        layer_sizes (list of int): list of output sizes of the graph GCN layers in the stack.
+        activations (list of str): list of activations applied to each GCN layer's output.
+        generator (GraphGenerator): an instance of :class:`GraphGenerator` class constructed on the graphs used for
+            training.
+        bias (bool, optional): toggles an optional bias in graph convolutional layers.
+        dropout (float, optional): dropout rate applied to input features of each GCN layer.
+        kernel_regularizer (str): normalization applied to the GCN layer kernels.
+    """
+
+    def __init__(
+        self, layer_sizes, activations, generator, bias=True, dropout=0.0, **kwargs
```

Comment:
Relatedly to the `kwargs`, but somewhat different (and hence a separate thread), this class is currently hard-coded to support only a single version of GCN, and thus has to pass through all the right arguments. I wonder if there's a more general form, similar to how #1001 generalised `GCNInfoMax` which was hard-coded to GCN to `DeepGraphInfoMax` that supports many kinds of layers (and can be generalised further). For instance, one could use GraphSAGE or some other inductive algorithms. What do you think?

Hypothetically, one could imagine having this be used something like:

```python
cluster_gcn = ClusterGCN(...)
graph_classification = GraphClassification(cluster_gcn)
```

Getting the generator support might be a little awkward, but I feel that we can get something working here. I could imagine something like `GraphGenerator(graphs, ClusterNodeGenerator, clusters=1, lam=3)` where basically the `GraphGenerator` manages a base generator for each graph or something.

We could still offer convenient ways to get a useful default approach.

Do you have any thoughts on this?

Reply:
Ultimately, this would be the best way to build a general graph classification model where the user has control of the type of graph convolution used. However, as I described in my comment about the need for the custom `GraphClassificationConvolution` layer, there are many implementation details for our existing layers and models that make it difficult to use as you suggest. The treatment of the batch dimension, the `out_indices` input tensor and the notion of a last layer for the existing models make the above solution a bit difficult to implement. I think if we resolve these issues then we can re-visit this.